designs/aws-launch-template-network-interfaces.md

```diff
@@ -0,0 +1,69 @@
+  securityGroupSelector:
+      securityGroup: sg-1 # users expect this to be used for all network interfaces
+  networkInterfaces:
+  - subnetSelector:
+        subnet: subnet-2 # users expect this to be used for this specific network interface
+    securityGroupSelector:
+        securityGroup: sg-2 # users expect this to be used for this specific network interface
+```
+having both options (`AWSNodeTemplate` level and networkInterface level) will complicate the API and will cause confusion for the users. It shall also complicate the logic on karpenter controller side.   
+In addition, removing the `AWSNodeTemplate`'s level subnet and security group selectors is a breaking change from API backward compatibility perspective.
+
+one more thing to consider is that subnets are bound to a single AZ. that's why the resolution happen in runtime instead of being part of the AWS launchTemplate (to avoid creating as many launch templates as the AZs in the zone).
+moving the resolution logic to the networkInterface level has the following complications: 
+1. more resolutions and interface allocations to do
+2. NetworkInterfaces can have conflicting requirements (e.g. different subnets in different AZs). which we need to handle in runtime.
+
+## networkInterfaces with specific IP addresses
+networkInterfaces allows specifying specific IP addresses for the network interface. This launchTemplate with specific IP addresses is not very useful for karpenter as it can not create many nodes without IP address collisions.
+
+## number of network interfaces per instance type
+the [maximum number of network interfaces a node can support](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-eni.html#AvailableIpPerENI) differs based on the instance-type. This creates dependency between the `Provisioner` resource and the `AWSNodeTemplate`.
+for example if the provisioner allows the `m5.large` and `m5.xlarge` instance types, while the `AWSNodeTemplate` specifies 4 network interfaces, karpenter controller needs to know that it can only provision `m5.xlarge` instances.
+
+## node drift
+there are requirements to reconcile nodes as they drift from the specs defined in the provisioner resource as defined [here](https://github.com/aws/karpenter/issues/1738) and [here](https://github.com/aws/karpenter/issues/1457).
+the topic of having multiple network interfaces per node adds to the complexity of the drift detection logic.
+
+# Proposed Solution:
+iterative approach where we start by exposing subset of the fields of networkInterfaces and then expand the support based on the demand.
+those fields are:
+```yaml
+  AssociateCarrierIpAddress: Boolean
+  AssociatePublicIpAddress: Boolean
+  DeleteOnTermination: Boolean
+  Description: String
+  DeviceIndex: Integer
+  InterfaceType: String
+  NetworkCardIndex: Integer
+```
+in addition, we will only support two networkInterfaces per node, this way we guarantee that the node will be able to run on all instance types.
```

Comment:
I think for this case, we just have to trust the user to craft a provisioner and AWSNodeTemplate that are compatible. The 2 limit is safe, but it's limiting if you've configured your provisioner to only have c6i.48xlarge instance types or something like that.

Reply:
Yeah, I thin runtime errors are okay for the complex validation failures having to do with AWS APIs parameter combinations. Anything that we can validate in a straight forward way on the CRD is good, but I'd stop short of any complicated validations and definitely stopping short at any AWS API calls to perform validation.